### ISSUE TITLE
"program.emit" produces wrong declaration file if "declaration" option is false or not specified

### ISSUE THREAD

**Issue by timocov** [microsoft/TypeScript#24002]
**TypeScript Version:** 2.9.0-dev.20180506

**Search Terms:** declaration, emitting.

**Code**

```ts
declare module ModuleName {}
declare var ModuleName: { prototype: any };
export { ModuleName };
```

Steps to reproduce:
1. Create `index.ts` with content above
1. Open `node` in folder with this file and installed `typescript` and type the following:

    ```js
    var ts = require('typescript');
    var program = ts.createProgram(['index.ts'], {});
    var res = program.emit(undefined, undefined, undefined, true);
    ```

**Expected behavior:**

Possible that calling `program.emit` with `emitOnlyDtsFiles = true` and without specifying `declaration: true` in the config should produce an error, but I am not sure that it is good solution.

**Actual behavior:**

The compiler produces `index.d.ts` file with only one line `export { ModuleName };` which is incorrect definitions (`res` is object with the following state: `{ emitSkipped: false, diagnostics: [], emittedFiles: undefined, sourceMaps: undefined }`).

For `2.8.3` version it produces

```ts
declare module ModuleName {
}
declare var ;
export { ModuleName };
```

which is broken declaration too.

If run from cmd `./node_modules/.bin/tsc --emitDeclarationOnly` it fails with `error TS5052: Option 'emitDeclarationOnly' cannot be specified without specifying option 'declaration'`.

**Comment by timocov** [comment]
Hey @RyanCavanaugh,

looks like you've changed the behavior in 3.7 and now if you haven't `declaration` option enabled you don't get any declaration file in output via call `program.emit` with `emitOnlyDtsFiles = true` (previously you were able to get them)? Is the correct solution right now to emit an error when `emitOnlyDtsFiles` is `true`, but `declaration` is disabled for call `program.emit` to fix this issue?

**Comment by RyanCavanaugh** [comment]
For API consumption purposes if the call is ill-formed (e.g. asking for only declaration files when `declaration` is false), the resulting behavior is undefined. I wouldn't really consider any of those options more correct than the other.

**Comment by timocov** [comment]
So, the correct behavior here is passing `declaration=true` in the compiler options whenever you want to generate declarations (the same as CLI requires it by throwing an error if the option is `false`)?

> the resulting behavior is undefined. I wouldn't really consider any of those options more correct than the other.

Looks like this issue cannot be solved? 😂

**Comment by RyanCavanaugh** [comment]
Behavior as-is is justifiable -- API calls that programmatically modify options have to either get config diagnostics or check that they're not doing something illegal
